zcm/julia/ZCM.jl

```diff
@@ -15,320 +59,276 @@ end
+    handler::F
+end
 
-            if (loadedRbuf.data != C_NULL)
-                instance.data = unsafe_wrap(Array, loadedRbuf.data, loadedRbuf.data_size);
-            else
-                instance.data = [];
-            end
-        end
+"""
+The Subscription type contains the SubscriptionOptions and also all of the
+various C pointers to the libuv handlers.
+"""
+immutable Subscription{T <: SubscriptionOptions}
+    options::T
+    c_handler::Ptr{Void}
+    uv_wrapper::Ptr{Native.UvSub}
+    uv_handler::Ptr{Void}
+    native_sub::Ptr{Native.Sub}
+end
 
-        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
+type Zcm
+    zcm::Ptr{Native.Zcm}
+    subscriptions::Vector{Subscription}
+
+    function Zcm(url::AbstractString = "")
+        pointer = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), url);
+        instance = new(pointer, Subscription[])
+        finalizer(instance, close)
+        return instance
+    end
+end
 
-        return instance;
+function close(zcm::Zcm)
+    if zcm.zcm != C_NULL
+        ccall(("zcm_destroy", "libzcm"), Void,
+              (Ptr{Native.Zcm},), zcm)
+        zcm.zcm = C_NULL
     end
 end
-export RecvBuf;
 
-type __zcm_handler
-    handler::Any;
-    usr::Any;
+# Defines the conversion when we pass this to a C function expecting a pointer
+unsafe_convert(::Type{Ptr{Native.Zcm}}, zcm::Zcm) = zcm.zcm
 
-    # handler must have the function or functor prototype :
-    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
-    __zcm_handler(handler, usr) = new(handler, usr);
+function good(zcm::Zcm)
+    (zcm.zcm != C_NULL) && (errno(zcm) == 0)
 end
 
-function __zcm_handler_wrapper(rbuf::Ptr{Native.RecvBuf}, channel::Cstring, usr_::Ptr{Void})
-    usr = unsafe_pointer_to_objref(usr_)::__zcm_handler;
-    usr.handler(RecvBuf(rbuf), unsafe_string(channel), usr.usr);
-    return nothing;
+function errno(zcm::Zcm)
+    ccall(("zcm_errno", "libzcm"), Cint, (Ptr{Native.Zcm},), zcm)
 end
 
-type Sub
-    sub::Ptr{Native.Sub};
-
-    # References to protect our callbacks from Julia's garbage collector
-    handler_jl::__zcm_handler;
-    handler_c ::Ptr{Void};
-    uv_handler::Ptr{Void};
-    uv_wrapper::Ptr{Native.UvSub};
-
-    Sub(sub::Ptr{Native.Sub}, handler_jl::__zcm_handler, handler_c::Ptr{Void},
-        uv_handler::Ptr{Void}, uv_wrapper::Ptr{Native.UvSub}) =
-        new(sub, handler_jl, handler_c, uv_handler, uv_wrapper);
+function strerror(zcm::Zcm)
+    val =  ccall(("zcm_strerror", "libzcm"), Cstring, (Ptr{Native.Zcm},), zcm)
+    if (val == C_NULL)
+        return "unable to get strerror"
+    else
+        return unsafe_string(val)
+    end
 end
 
-type Zcm
-    zcm::Ptr{Native.Zcm};
+function handler_wrapper(rbuf::Native.RecvBuf, channelbytes::Cstring, opts::SubscriptionOptions)
+    channel = unsafe_string(channelbytes)
+    msgdata = unsafe_wrap(Vector{UInt8}, rbuf.data, rbuf.data_size)
+    msg = decode(opts.msgtype, msgdata)
```

Comment:
Let's push this decode out into the closure you wrap the subscribe with. that way we can support untyped subscriptions. Would like to be able to support a `subscribe(".*")` with no type. Will also still need the rbuf passed into the handler. We use rbuf all over

Reply:
Interesting. How is the rbuf used in user code? It looks like the Python and Node bindings don't pass it to the handler.

Reply:
If they don't that's definitely not right haha. We use the recv utime which is populated by the transport. It's the time the message was received.

Reply:
Ok, sounds good. I agree that it makes sense to support untyped subscriptions, so I'll make that possible too.

Reply:
Ok, I've removed the msgtype argument from subscribe in 350c3bf. I also added a `typed_handler(handler, msgtype)` function which constructs the appropriate closure, since I was getting tired of writing:

```
(rbuf, channel, msgdata, args...) -> (rbuf, channel, decode(example_t, msgdata), args...)
```

instead you can do:

```
typed_callback(handler, example_t)
```